contrib/externalMatchfinder/main.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) Yann Collet, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under both the BSD-style license (found in the
+ * LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ * in the COPYING file in the root directory of this source tree).
+ * You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define ZSTD_STATIC_LINKING_ONLY
+#include "zstd.h"
+#include "zstd_errors.h"
+#include "matchfinder.h" // simpleExternalMatchFinder
+
+int main(int argc, char *argv[]) {
+    size_t res;
+
+    if (argc != 2) {
+        printf("Usage: exampleMatchfinder <file>\n");
+        return 1;
+    }
+
+    ZSTD_CCtx* zc = ZSTD_createCCtx();
+
+    int simpleExternalMatchState = 0xdeadbeef;
+
+    // Here is the crucial bit of code!
+    ZSTD_refExternalMatchFinder(
+        zc,
+        &simpleExternalMatchState,
+        simpleExternalMatchFinder
+    );
+
+    res = ZSTD_CCtx_setParameter(zc, ZSTD_c_enableMatchFinderFallback, 1);
+
+    if (ZSTD_isError(res)) {
+        printf("ERROR: %s\n", ZSTD_getErrorName(res));
+        return 1;
+    }
+
+    FILE *f = fopen(argv[1], "rb");
+    fseek(f, 0, SEEK_END);
+    long const srcSize = ftell(f);
+    fseek(f, 0, SEEK_SET);
+
+    char *src = malloc(srcSize + 1);
+    fread(src, srcSize, 1, f);
+    fclose(f);
+
+    size_t const dstSize = ZSTD_compressBound(srcSize);
+    char *dst = malloc(dstSize);
+
+    size_t const cSize = ZSTD_compress2(zc, dst, dstSize, src, srcSize);
+
+    if (ZSTD_isError(cSize)) {
+        printf("ERROR: %s\n", ZSTD_getErrorName(cSize));
+        return 1;
+    }
+
+    char *val = malloc(srcSize);
+    res = ZSTD_decompress(val, srcSize, dst, cSize);
+
+    ZSTD_freeCCtx(zc);
```

Comment:
minor: it feels strange to free `zc` here, in the middle of decompression.
I would recommend either to free it at end of compression (just 2 lines above),
or at end of function